lib/plugins/publicHealthcheckPlugin.ts

```diff
@@ -4,42 +4,61 @@ import fp from 'fastify-plugin'
 
 export interface PublicHealthcheckPluginOptions {
   responsePayload?: Record<string, unknown>
+  url?: string
+  logLevel?: 'fatal' | 'error' | 'warn' | 'info' | 'debug' | 'trace' | 'silent'
   healthChecks: readonly HealthCheck[]
 }
 
-export type HealthCheck = (app: FastifyInstance) => Promise<Either<Error, true>>
+export type HealthChecker = (app: FastifyInstance) => Promise<Either<Error, true>>
+
+export type HealthCheck = {
+  isMandatory: boolean
+  checker: HealthChecker
+}
 
 function plugin(app: FastifyInstance, opts: PublicHealthcheckPluginOptions, done: () => void) {
   const responsePayload = opts.responsePayload ?? {}
   app.route({
-    url: '/',
+    url: opts.url ?? '/health',
     method: 'GET',
-    logLevel: 'debug',
+    logLevel: opts.logLevel ?? 'info',
     schema: {
       // hide route from swagger plugins
       // @ts-expect-error
       hide: true,
     },
     handler: async (_, reply) => {
       let isHealthy = true
+      let isPartiallyHealthy = false
       if (opts.healthChecks.length) {
         const results = await Promise.all(
           opts.healthChecks.map((healthcheck) => {
-            return healthcheck(app)
+            return healthcheck.checker(app)
           }),
         )
         if (
-          results.find((entry) => {
-            return !!entry.error
+          results.find((entry, index) => {
+            return entry.error && opts.healthChecks[index].isMandatory
           })
         ) {
           isHealthy = false
         }
+
+        if (isHealthy) {
+          if (
+            results.find((entry, index) => {
+              return entry.error && !opts.healthChecks[index].isMandatory
+            })
+          ) {
+            isHealthy = false
+            isPartiallyHealthy = true
```

Comment:
Although tests looks correct, this logic is a bit tricky to follow. Only if `isHealthy === true` then `isPartiallyHealthy` can become `true`? 🤔

Could this be done in a single loop, and then
- If any healthcheck passes then set `isPartiallyHealthy` to `true`.
- If any healthcheck fails then set `isHealthy` to `false`
?

Reply:
Agreed that this is very hard to follow. Here is a rewrite:

```ts
    handler: async (_, reply) => {
      const mandatoryChecks = opts.healthChecks.filter(({ isMandatory }) => isMandatory)
      const optionalChecks = opts.healthChecks.filter(({ isMandatory }) => !isMandatory)

      const [
        mandatoryCheckResults,
        optionalCheckResults
      ] = await Promise.all([
        Promise.all(mandatoryChecks.map((healthCheck) => healthCheck.checker(app))),
        Promise.all(optionalChecks.map((healthCheck) => healthCheck.checker(app))),
      ])

      // If any mandatory check has an error, then the health check has failed
      if(mandatoryCheckResults.some((result) => !!result.error)) {
        return reply.status(500).send({
          ...responsePayload,
          heartbeat: 'FAIL'
        })
      }

      // If any optional check has an error, then we are partially healthy
      if(optionalCheckResults.some((result) => !!result.error)) {
        return reply.status(200).send({
          ...responsePayload,
          heartbeat: 'PARTIALL_HEALTHY'
        })
      }

      // No checks had an error. Note this includes the case that there were no checks at all.
      return reply.status(200).send({
        ...responsePayload,
        heartbeat: 'HEALTHY'
      })
    },

```